File: src/relationship/runtime/sensorEvents.ts
/**
 * Zod contract for macOS sensor NDJSON events.
 *
 * Field names and event shapes mirror the Swift emitter in
 * `swift/FriendyMacOSSensor/Sources/FriendyMacOSSensor/SensorEvents.swift` so
 * TypeScript and the native sensor stay in parity without a shared codegen step.
 *
 * Privacy: contact payloads carry hashed methods and redacted hints only. Raw
 * phone numbers and email addresses are rejected before Zod parsing.
 */
import { z } from "zod";

/** Stable sensor identifier shared with the Swift emitter. */
export const MACOS_SENSOR_NAME = "macos_contacts_calendar";
/** Schema version enforced on every NDJSON line before type-specific parsing. */
export const MACOS_SENSOR_SCHEMA_VERSION = 1;

const permissionStatusSchema = z.enum(["authorized", "denied", "restricted", "notDetermined", "unavailable"]);

const commonEventSchema = z.object({
  schemaVersion: z.literal(MACOS_SENSOR_SCHEMA_VERSION),
  eventId: z.string().min(1),
  type: z.string().min(1),
  sensorName: z.literal(MACOS_SENSOR_NAME),
  sensorVersion: z.string().min(1),
  runId: z.string().min(1),
  deviceId: z.string().min(1),
  emittedAt: z.string().min(1)
});

const contactMethodHintSchema = z.object({
  last4: z.string().min(1).optional(),
  domain: z.string().min(1).optional(),
  label: z.string().min(1).optional()
});

const contactSchema = z.object({
  stableId: z.string().min(1),
  unifiedStableId: z.string().min(1).optional(),
  containerId: z.string().min(1).optional(),
  displayName: z.string().min(1),
  phoneNumberHashes: z.array(z.string().min(1)).default([]),
  phoneNumberHints: z.array(contactMethodHintSchema).default([]),
  emailHashes: z.array(z.string().min(1)).default([]),
  emailHints: z.array(contactMethodHintSchema).default([])
});

const calendarMatchSchema = z.object({
  eventIdentifier: z.string().min(1).optional(),
  calendarIdentifier: z.string().min(1).optional(),
  title: z.string().min(1),
  startsAt: z.string().min(1),
  endsAt: z.string().min(1),
  location: z.string().optional(),
  calendarSource: z.string().min(1),
  calendarTitle: z.string().min(1),
  isAllDay: z.boolean(),
  attendeeCount: z.number().int().nonnegative().default(0),
  availability: z.string().optional(),
  status: z.string().optional(),
  isRecurring: z.boolean().default(false)
});

const calendarQuerySchema = z.object({
  startsAt: z.string().min(1),
  endsAt: z.string().min(1),
  resultCountBeforeLimit: z.number().int().nonnegative(),
  permissionStatus: permissionStatusSchema,
  errorCode: z.string().min(1).optional()
});

const readyEventSchema = commonEventSchema.extend({
  type: z.literal("ready"),
  contactsPermissionStatus: permissionStatusSchema,
  calendarPermissionStatus: permissionStatusSchema,
  baselineCreated: z.boolean()
});

const contactAddedEventSchema = commonEventSchema.extend({
  type: z.literal("contact_added"),
  observedAt: z.string().min(1),
  idempotencyKey: z.string().min(1),
  historyBatchId: z.string().min(1),
  historyBatchIndex: z.number().int().nonnegative(),
  historyBatchSize: z.number().int().positive(),
  historyTokenBeforeRef: z.string().min(1),
  historyTokenAfterRef: z.string().min(1),
  detectedAt: z.string().min(1),
  contact: contactSchema,
  calendarQuery: calendarQuerySchema,
  calendarMatches: z.array(calendarMatchSchema).default([])
});

const historyBatchCompleteEventSchema = commonEventSchema.extend({
  type: z.literal("history_batch_complete"),
  historyBatchId: z.string().min(1),
  contactEventIds: z.array(z.string().min(1)),
  ackPath: z.string().min(1)
});

const historyResetEventSchema = commonEventSchema.extend({
  type: z.literal("history_reset"),
  idempotencyKey: z.string().min(1),
  reason: z.string().min(1),
  detectedAt: z.string().min(1)
});

const permissionErrorEventSchema = commonEventSchema.extend({
  type: z.literal("permission_error"),
  idempotencyKey: z.string().min(1),
  code: z.string().min(1),
  message: z.string().min(1),
  retryable: z.boolean()
});

const fatalErrorEventSchema = commonEventSchema.extend({
  type: z.literal("fatal_error"),
  idempotencyKey: z.string().min(1),
  code: z.string().min(1),
  message: z.string().min(1),
  retryable: z.boolean()
});

const sensorEventSchema = z.discriminatedUnion("type", [
  readyEventSchema,
  contactAddedEventSchema,
  historyBatchCompleteEventSchema,
  historyResetEventSchema,
  permissionErrorEventSchema,
  fatalErrorEventSchema
]);

export type MacosPermissionStatus = z.infer<typeof permissionStatusSchema>;
/** Contact payload with hashed methods and redacted hints; never raw phone/email strings. */
export type MacosContactPayload = z.infer<typeof contactSchema>;
export type MacosCalendarMatch = z.infer<typeof calendarMatchSchema>;
/** Discriminated union of all macOS sensor NDJSON event types. */
export type MacosSensorEvent = z.infer<typeof sensorEventSchema>;
export type MacosContactAddedEvent = Extract<MacosSensorEvent, { type: "contact_added" }>;

/**
 * Parses one NDJSON line from the macOS sensor stdout stream.
 *
 * Validates common contract fields, rejects forbidden raw contact method keys,
 * then runs the full Zod schema for the event `type`.
 */
export function parseSensorEventLine(line: string): MacosSensorEvent {
  const parsed = parseJsonObject(line);
  assertCommonContract(parsed);
  assertNoRawContactMethods(parsed);

  const result = sensorEventSchema.safeParse(parsed);
  if (!result.success) {
    throw new Error(`Invalid macOS sensor event: ${z.prettifyError(result.error)}`);
  }

  return result.data;
}

function parseJsonObject(line: string): Record<string, unknown> {
  try {
    const parsed = JSON.parse(line) as unknown;
    if (!isRecord(parsed)) {
      throw new Error("sensor event must be a JSON object");
    }
    return parsed;
  } catch (error) {
    if (error instanceof SyntaxError) {
      throw new Error(`Malformed sensor JSON: ${error.message}`);
    }
    throw error;
  }
}

function assertCommonContract(payload: Record<string, unknown>): void {
  if (payload.schemaVersion !== MACOS_SENSOR_SCHEMA_VERSION) {
    throw new Error(`Invalid macOS sensor event: schemaVersion must be ${MACOS_SENSOR_SCHEMA_VERSION}`);
  }

  if (payload.sensorName !== MACOS_SENSOR_NAME) {
    throw new Error(`Invalid macOS sensor event: sensorName must be ${MACOS_SENSOR_NAME}`);
  }
}

/** Rejects Swift regressions that leak raw phone/email fields before Zod runs. */
function assertNoRawContactMethods(payload: Record<string, unknown>): void {
  if (payload.type !== "contact_added" || !isRecord(payload.contact)) {
    return;
  }

  const forbiddenKeys = ["phoneNumbers", "emails", "phoneNumber", "email"];
  const present = forbiddenKeys.filter((key) => Object.prototype.hasOwnProperty.call(payload.contact, key));
  if (present.length > 0) {
    throw new Error(`Invalid macOS sensor event: raw contact method fields are not allowed (${present.join(", ")})`);
  }
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

File: swift/FriendyMacOSSensor/Sources/FriendyMacOSSensor/SensorEvents.swift
// FriendyMacOSSensor — NDJSON event builders and stdout emission.
// Schema parity: each printed line must validate against `src/relationship/runtime/sensorEvents.ts`
// (`MACOS_SENSOR_SCHEMA_VERSION`, `MACOS_SENSOR_NAME`, discriminated `type`, no raw phone/email fields).

import Foundation

/// Sensor identifier shared with the TypeScript runtime (`macos_contacts_calendar`).
let friendySensorName = "macos_contacts_calendar"
let friendySensorVersion = "0.1.0"

/// When set, each NDJSON line is appended here (used when macOS launches the sensor via `.app` bundle).
var sensorEventLogPath: String?

/// Per-run identity fields included on every emitted event.
struct SensorIdentity {
    let runId: String
    let deviceId: String

    static func current() -> SensorIdentity {
        SensorIdentity(
            runId: "sensor_run_\(UUID().uuidString)",
            deviceId: Host.current().localizedName ?? "mac_local"
        )
    }
}

/// Encodes one JSON object and prints a single NDJSON line to stdout (flushed immediately).
func emitSensorEvent(_ event: [String: Any]) {
    guard JSONSerialization.isValidJSONObject(event),
          let data = try? JSONSerialization.data(withJSONObject: event, options: []),
          let line = String(data: data, encoding: .utf8) else {
        fputs("Failed to encode sensor event\n", stderr)
        exit(1)
    }

    print(line)
    fflush(stdout)

    if let logPath = sensorEventLogPath, let data = (line + "\n").data(using: .utf8) {
        if FileManager.default.fileExists(atPath: logPath),
           let handle = FileHandle(forWritingAtPath: logPath) {
            handle.seekToEndOfFile()
            handle.write(data)
            handle.closeFile()
        } else if !FileManager.default.createFile(atPath: logPath, contents: data) {
            fputs("Failed to write sensor event log at \(logPath)\n", stderr)
        }
    }
}

/// Common envelope: schemaVersion, eventId, type, sensorName, sensorVersion, runId, deviceId, emittedAt.
func commonSensorEvent(_ type: String, identity: SensorIdentity, now: Date = Date()) -> [String: Any] {
    [
        "schemaVersion": 1,
        "eventId": "sensor_evt_\(UUID().uuidString)",
        "type": type,
        "sensorName": friendySensorName,
        "sensorVersion": friendySensorVersion,
        "runId": identity.runId,
        "deviceId": identity.deviceId,
        "emittedAt": ISO8601DateFormatter().string(from: now)
    ]
}

func fatalSensorEvent(code: String, message: String, identity: SensorIdentity) -> [String: Any] {
    var event = commonSensorEvent("fatal_error", identity: identity)
    event["idempotencyKey"] = "fatal_error:\(identity.deviceId):\(identity.runId):\(code)"
    event["code"] = code
    event["message"] = message
    event["retryable"] = false
    return event
}

func readySensorEvent(identity: SensorIdentity, baselineCreated: Bool, calendarPermissionStatus: String) -> [String: Any] {
    var event = commonSensorEvent("ready", identity: identity)
    event["contactsPermissionStatus"] = "authorized"
    event["calendarPermissionStatus"] = calendarPermissionStatus
    event["baselineCreated"] = baselineCreated
    return event
}

func contactAddedFixtureEvent(identity: SensorIdentity) -> [String: Any] {
    contactAddedEvent(
        identity: identity,
        eventId: "sensor_evt_fixture_contact_1",
        idempotencyKey: "contacts:\(identity.deviceId):fixture-contact-1:add",
        historyBatchId: "history_batch_fixture_1",
        historyBatchIndex: 0,
        historyBatchSize: 1,
        historyTokenBeforeRef: "outbox:history_batch_fixture_1:before",
        historyTokenAfterRef: "outbox:history_batch_fixture_1:after",
        detectedAt: "2026-05-21T20:30:00-07:00",
        observedAt: "2026-05-21T18:36:50Z",
        contact: [
            "stableId": "fixture-contact-1",
            "unifiedStableId": "fixture-contact-1",
            "containerId": "fixture-container",
            "displayName": "Maya",
            "phoneNumberHashes": ["sha256:fixture-phone"],
            "phoneNumberHints": [["last4": "4567", "label": "mobile"]],
            "emailHashes": ["sha256:fixture-email"],
            "emailHints": [["domain": "example.com", "label": "work"]]
        ],
        calendarQuery: [
            "startsAt": "2026-05-21T16:30:00-07:00",
            "endsAt": "2026-05-21T21:30:00-07:00",
            "resultCountBeforeLimit": 1,
            "permissionStatus": "authorized"
        ],
        calendarMatches: [
            [
                "eventIdentifier": "fixture-event-photon-dinner",
                "calendarIdentifier": "fixture-calendar-work",
                "title": "Photon Residency Dinner",
                "startsAt": "2026-05-21T18:00:00-07:00",
                "endsAt": "2026-05-21T21:00:00-07:00",
                "location": "San Francisco",
                "calendarSource": "iCloud",
                "calendarTitle": "Work",
                "isAllDay": false,
                "attendeeCount": 12,
                "availability": "busy",
                "status": "confirmed",
                "isRecurring": false
            ]
        ]
    )
}

/// `contact_added` payload; contact uses hashes/hints only (never raw phoneNumbers or emails).
func contactAddedEvent(
    identity: SensorIdentity,
    eventId: String,
    idempotencyKey: String,
    historyBatchId: String,
    historyBatchIndex: Int,
    historyBatchSize: Int,
    historyTokenBeforeRef: String,
    historyTokenAfterRef: String,
    detectedAt: String,
    observedAt: String,
    contact: [String: Any],
    calendarQuery: [String: Any],
    calendarMatches: [[String: Any]]
) -> [String: Any] {
    var event = commonSensorEvent("contact_added", identity: identity)
    event["eventId"] = eventId
    event["observedAt"] = observedAt
    event["idempotencyKey"] = idempotencyKey
    event["historyBatchId"] = historyBatchId
    event["historyBatchIndex"] = historyBatchIndex
    event["historyBatchSize"] = historyBatchSize
    event["historyTokenBeforeRef"] = historyTokenBeforeRef
    event["historyTokenAfterRef"] = historyTokenAfterRef
    event["detectedAt"] = detectedAt
    event["contact"] = contact
    event["calendarQuery"] = calendarQuery
    event["calendarMatches"] = calendarMatches
    return event
}

func contactPendingEvent(
    identity: SensorIdentity,
    reason: String,
    pendingContactCount: Int,
    readyContactCount: Int? = nil,
    nextCheckInSeconds: Int? = nil
) -> [String: Any] {
    var event = commonSensorEvent("contact_pending", identity: identity)
    event["reason"] = reason
    event["pendingContactCount"] = pendingContactCount
    if let readyContactCount {
        event["readyContactCount"] = readyContactCount
    }
    if let nextCheckInSeconds {
        event["nextCheckInSeconds"] = nextCheckInSeconds
    }
    return event
}

func sensorDiagnosticEvent(
    identity: SensorIdentity,
    code: String,
    pendingContactCount: Int? = nil,
    nextCheckInSeconds: Int? = nil
) -> [String: Any] {
    var event = commonSensorEvent("sensor_diagnostic", identity: identity)
    event["code"] = code
    if let pendingContactCount {
        event["pendingContactCount"] = pendingContactCount
    }
    if let nextCheckInSeconds {
        event["nextCheckInSeconds"] = nextCheckInSeconds
    }
    return event
}

func historyBatchCompleteFixtureEvent(identity: SensorIdentity) -> [String: Any] {
    historyBatchCompleteEvent(
        identity: identity,
        historyBatchId: "history_batch_fixture_1",
        contactEventIds: ["sensor_evt_fixture_contact_1"],
        ackPath: ".friendy/macos-sensor-state/acks/history_batch_fixture_1.ack"
    )
}

func historyBatchCompleteEvent(identity: SensorIdentity, historyBatchId: String, contactEventIds: [String], ackPath: String) -> [String: Any] {
    var event = commonSensorEvent("history_batch_complete", identity: identity)
    event["historyBatchId"] = historyBatchId
    event["contactEventIds"] = contactEventIds
    event["ackPath"] = ackPath
    return event
}

func historyResetEvent(identity: SensorIdentity, reason: String, detectedAt: String) -> [String: Any] {
    var event = commonSensorEvent("history_reset", identity: identity)
    event["idempotencyKey"] = "history_reset:\(identity.deviceId):\(identity.runId):\(reason):\(detectedAt)"
    event["reason"] = reason
    event["detectedAt"] = detectedAt
    return event
}

/*
The fixture event keeps the contract visible in source while avoiding raw phoneNumbers/emails:
[
    "contact": [
        "stableId": "fixture-contact-1",
        "unifiedStableId": "fixture-contact-1",
        "containerId": "fixture-container",
        "displayName": "Maya",
        "phoneNumberHashes": ["sha256:fixture-phone"],
        "phoneNumberHints": [["last4": "4567", "label": "mobile"]],
        "emailHashes": ["sha256:fixture-email"],
        "emailHints": [["domain": "example.com", "label": "work"]]
    ]
    event["calendarQuery"] = [
        "startsAt": "2026-05-21T16:30:00-07:00",
        "endsAt": "2026-05-21T21:30:00-07:00",
        "resultCountBeforeLimit": 1,
        "permissionStatus": "authorized"
    ]
    event["calendarMatches"] = [
        [
            "eventIdentifier": "fixture-event-photon-dinner",
            "calendarIdentifier": "fixture-calendar-work",
            "title": "Photon Residency Dinner",
            "startsAt": "2026-05-21T18:00:00-07:00",
            "endsAt": "2026-05-21T21:00:00-07:00",
            "location": "San Francisco",
            "calendarSource": "iCloud",
            "calendarTitle": "Work",
            "isAllDay": false,
            "attendeeCount": 12,
            "availability": "busy",
            "status": "confirmed",
            "isRecurring": false
        ]
    ]
]
*/
